### PR TITLE
Refactor everything into a class [READY?]

### PR DESCRIPTION
**Pull request by mvdoc**
I started refactoring everything into a class, still working on it -- not ready yet to merge.

Please have a look and let me know if you spot something terrible that needs to change. I ported most of the tests.

### CODE REVIEW

File: reprclust/cluster_methods.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+        Clusters the data. args and kwargs depend on the current method.
+    predict(*args, **kwargs)
+        Predicts clustering on new data based on previous cluster solution.
+        args and kwargs depend on the current method.
+    get_clusters(k)
+        Returns the cluster solution for k clusters.
+    get_predicted_clusters(k)
+        Returns the predicted cluster solution for k clusters.
+    """
+    def __init__(self, clustering_method):
```
Comment (yarikoptic) [inline reprclust/cluster_methods.py]: no need to pass the specific implementation  function here. Derived classes will call necessary function themselves.  Generally you don't want to mask out parameters for the methods of the parent class -- you just want to extend its API (e.g. with new parameters like parameters for the specific clustering implementation)

File: reprclust/cluster_methods.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                args.append("{0}={1}".format(key, repr(value)))
+        if args:
+            args = ", ".join(args)
+        else:
+            args = ""
+
+        return self.__class__.__name__ + '({0})'.format(args)
+
+    @property
+    def run(self):
```
Comment (yarikoptic) [inline reprclust/cluster_methods.py]: needs to be reconsidered -- joblib memoization (?)

File: reprclust/cluster_methods.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        super(KMeansClusterMethod, self).__init__(KMeans)
+        self._args = args
+        self._kwargs = kwargs
+        self._method_output = None
+        self._methods = {}
+
+    def cluster(self, k):
+        if k not in self._clusters:
+            kmeans = self.method(n_clusters=k, *self._args, **self._kwargs)
+            self._methods[k] = kmeans.fit(self.data)
```
Comment (yarikoptic) [inline reprclust/cluster_methods.py]: may be not _methods, but _models or smth like that

File: reprclust/cluster_methods.py [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+            self._clusters[k] = kmeans.predict(self.data)
+
+    def get_method(self, k):
+        """Return the fitted KMeans with k components"""
+        if k in self._methods:
+            return self._methods[k]
```
Comment (yarikoptic) [inline reprclust/cluster_methods.py]: return ```self._methods.get(k, None)```

File: reprclust/reproducibility.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    def ks(self):
+        return self._ks
+
+    def get_ks_run(self):
+        """Return the ks for the run of the clustering algorithm"""
+        metric = self.get_metric_names()[0]
+        return self.scores[metric][0].astype(int)
+
+    def get_metric_score(self, metric):
+        return self.scores[metric][1]
```
Comment (yarikoptic) [inline reprclust/reproducibility.py]: why ```[1]```?

File: reprclust/reproducibility.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        return self.__class__.__name__ + '({0})'.format(args)
+
+    def run(self, n_jobs=1, verbose=52):
+        parallel = Parallel(n_jobs=n_jobs, verbose=verbose)
+        fold = delayed(_run_fold)
+        result = parallel(fold(self, train, test)
+                          for train, test in self._splitter)
+        # store everything together now
+        for metrics in result[0]:
+            self.scores[metrics] = np.hstack((res[metrics] for res in result))
```
Comment (yarikoptic) [inline reprclust/reproducibility.py]: figure out correct plural/single endings

File: reprclust/cluster_methods.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+        # we'll store the predicted clusters here
+        self._predicted = {}
+        self.data = None
+
+        self._args = None
+        self._kwargs = None
+
+    def __call__(self, data):
+        """Just store the input data"""
+        self.data = data
```
Comment (mvdoc) [inline reprclust/cluster_methods.py]: do not store data

File: reprclust/cluster_metrics.py [hunk 8]
```diff
@@ -0,0 +1,154 @@
+# -*- coding: utf-8 -*-
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the module for the
+#   copyright and license terms.
+#
+### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""
+Module containing metrics to compare cluster solutions
```
Comment (mvdoc) [inline reprclust/cluster_metrics.py]: make scores classes

File: reprclust/reproducibility.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        for metric in self._cluster_metrics:
+            result_fold[metric.__name__][1, i_k] = \
+                metric(predicted_label, test_label)
+            if self._ground_truth is not None:
+                result_fold[metric.__name__ + '_gt'][1, i_k] = \
+                    metric(predicted_label, self._ground_truth)
+    return result_fold
+
+
+class Reproducibility(object):
```
Comment (mvdoc) [inline reprclust/reproducibility.py]: doesn't need to be class

File: reprclust/reproducibility.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        result_fold[metric.__name__] = \
+            np.vstack((self._ks, np.zeros(len(self._ks))))
+        if self._ground_truth is not None:
+            result_fold[metric.__name__ + '_gt'] = \
+                np.vstack((self._ks, np.zeros(len(self._ks))))
+
+    # Step 1. Clustering on training/test set and prediction
+    for k in self._ks:
+        # cluster on training set
+        cm_train.cluster(k)
```
Comment (mvdoc) [inline reprclust/reproducibility.py]: add option `fully_train` for cases like hierarchical clustering, which will be ignored by other methods

**Comment by mvdoc** [comment]
now the code is much simpler, and I also pushed the notebook with analysis on the simulation, and everything looks fine (i.e., we get the same results).

**Comment by mvdoc** [comment]
Merging

**Pull request merged**